tests/test_bootstrap.py

```diff
@@ -825,3 +829,131 @@ async def test_bootstrap_empty_integrations(
     """Test setting up an empty integrations does not raise."""
     await bootstrap.async_setup_multi_components(hass, set(), {})
     await hass.async_block_till_done()
+
+
+@pytest.mark.parametrize("integration", ["mqtt_eventstream", "mqtt_statestream"])
+@pytest.mark.parametrize("load_registries", [False])
+async def test_bootstrap_dependencies(
+    hass: HomeAssistant,
+    caplog: pytest.LogCaptureFixture,
+    integration: str,
+) -> None:
+    """Test dependencies are set up correctly,."""
+
+    # Prepare MQTT config entry
+    @HANDLERS.register("mqtt")
+    class MockConfigFlow:
+        """Mock the MQTT config flow."""
+
+        VERSION = 1
+
+    entry = MockConfigEntry(domain="mqtt", data={"broker": "test-broker"})
+    entry.add_to_hass(hass)
+
+    calls: list[str] = []
+    assertions: list[bool] = []
+
+    async def async_mqtt_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        """Assert the mqtt config entry was set up."""
+        calls.append("mqtt")
+        # assert the integration is not yet set up
+        assertions.append(hass.data["setup_done"][integration].is_set() is False)
+        assertions.append(
+            all(
+                dependency in hass.config.components
+                for dependency in integrations[integration]["dependencies"]
+            )
+        )
+        assertions.append(integration not in hass.config.components)
+        return True
+
+    async def async_integration_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+        """Assert the mqtt config entry was set up."""
+        calls.append(integration)
+        # assert mqtt was already set up
+        assertions.append(
+            "mqtt" not in hass.data["setup_done"]
+            or hass.data["setup_done"]["mqtt"].is_set()
+        )
+        assertions.append("mqtt" in hass.config.components)
+        return True
+
+    mqtt_integration = mock_integration(
+        hass,
+        MockModule(
+            "mqtt",
+            async_setup_entry=async_mqtt_setup_entry,
+            dependencies=["file_upload", "http"],
+        ),
+    )
+    mqtt_integration._import_platform = Mock()
```

Comment:
Why is this mock needed?

Reply:
It was a follow up on a comment of @bdraco as we do not want to make core tests to depend on non core components

Reply:
I'd add a comment to explain it.

Reply:
Added to https://github.com/home-assistant/core/pull/91803